client/trino-jdbc/src/test/java/io/trino/jdbc/TestJdbcConnection.java

```diff
@@ -482,9 +482,7 @@ public void testConcurrentCancellationOnStatementClose()
             statement.close();
 
             // verify that the query was cancelled
-            assertThatThrownBy(future::get)
-                    .hasCauseInstanceOf(SQLException.class)
-                    .hasStackTraceContaining("Error executing query");
```

Comment:
Maybe instead, change
```
future = executor.submit(() -> statement.execute(sql));
```

above

to something that would fire & consume the results

Reply:
> Maybe instead,

Instead does not work. As `future::get` may fail with other exceptions (including interrupted exception or "Error fetching results").

So instead changed the the mentioned lambda additionally.

Reply:
why could it be interrupted? 
did you experience "Error fetching results"? query cancellation should result in this, could it?

Reply:
```
java.util.concurrent.ExecutionException: java.sql.SQLException: Error fetching results
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:62)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:877)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1306)
	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1178)
	at io.trino.jdbc.TestJdbcConnection.lambda$testConcurrentCancellationOnConnectionClose$8(TestJdbcConnection.java:529)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:406)
	at io.trino.jdbc.TestJdbcConnection.testConcurrentCancellationOnConnectionClose(TestJdbcConnection.java:529)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeoutWithNoExecutor(MethodInvocationHelper.java:251)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeout(MethodInvocationHelper.java:240)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:651)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.sql.SQLException: Error fetching results
	at io.trino.jdbc.AbstractTrinoResultSet.next(AbstractTrinoResultSet.java:228)
	at io.trino.jdbc.TrinoResultSet.next(TrinoResultSet.java:43)
	at io.trino.jdbc.TestJdbcConnection.lambda$testConcurrentCancellationOnConnectionClose$5(TestJdbcConnection.java:513)
	... 4 more
Caused by: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2396)
	at io.trino.jdbc.TrinoResultSet$AsyncIterator.cancel(TrinoResultSet.java:140)
	at io.trino.jdbc.TrinoResultSet.close(TrinoResultSet.java:86)
	at io.trino.jdbc.TrinoStatement.closeResultSet(TrinoStatement.java:645)
	at io.trino.jdbc.TrinoStatement.close(TrinoStatement.java:100)
	at io.trino.jdbc.TrinoConnection.close(TrinoConnection.java:241)
	at io.trino.jdbc.TestJdbcConnection.testConcurrentCancellationOnConnectionClose(TestJdbcConnection.java:526)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeoutWithNoExecutor(MethodInvocationHelper.java:251)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeout(MethodInvocationHelper.java:240)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:651)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
	... 4 more
```

This looks reasonable as it is using `io.trino.jdbc.TrinoResultSet.ResultsPageIterator` which is cancelled.

Reply:
```
java.lang.InterruptedException
	at java.base/java.util.concurrent.FutureTask.awaitDone(FutureTask.java:418)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:190)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:62)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:877)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1306)
	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1178)
	at io.trino.jdbc.TestJdbcConnection.lambda$testConcurrentCancellationOnConnectionClose$8(TestJdbcConnection.java:529)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:406)
	at io.trino.jdbc.TestJdbcConnection.testConcurrentCancellationOnConnectionClose(TestJdbcConnection.java:529)
	at jdk.internal.reflect.GeneratedMethodAccessor2430.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeoutWithNoExecutor(MethodInvocationHelper.java:251)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeout(MethodInvocationHelper.java:240)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:651)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
```

Current thread, that was waiting on `future::get` got interrupted. Not sure why.